elements/castbar.lua

```diff
@@ -83,462 +85,289 @@ A default texture will be applied to the StatusBar and Texture widgets if they d
-		return element:PostCastInterruptible(unit)
-	end
-end
+	local delta
+	if(element.casting) then
+		delta = startTime - element.startTime
 
-local function UNIT_SPELLCAST_NOT_INTERRUPTIBLE(self, event, unit)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+		element.duration = GetTime() - startTime
+	else
+		delta = element.startTime - startTime
 
-	local element = self.Castbar
-	local shield = element.Shield
-	if(shield) then
-		shield:Show()
+		element.duration = endTime - GetTime()
 	end
 
-	element.notInterruptible = true
-
-	--[[ Callback: Castbar:PostCastNotInterruptible(unit)
-	Called after the element has been updated when a spell cast has become non-interruptible.
-
-	* self - the Castbar widget
-	* unit - unit for which the update has been triggered (string)
-	--]]
-	if(element.PostCastNotInterruptible) then
-		return element:PostCastNotInterruptible(unit)
+	if(delta < 0) then
+		delta = 0
 	end
-end
-
-local function UNIT_SPELLCAST_DELAYED(self, event, unit)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
-
-	local element = self.Castbar
-	local name, _, _, startTime = UnitCastingInfo(unit)
-	if(not startTime or not element:IsShown()) then return end
-
-	local duration = GetTime() - (startTime / 1000)
-	if(duration < 0) then duration = 0 end
 
-	element.delay = element.delay + element.duration - duration
-	element.duration = duration
+	element.max = endTime - startTime
+	element.startTime = startTime
+	element.delay = element.delay + delta
 
-	element:SetValue(duration)
+	element:SetMinMaxValues(0, element.max)
+	element:SetValue(element.duration)
 
-	--[[ Callback: Castbar:PostCastDelayed(unit, name)
-	Called after the element has been updated when a spell cast has been delayed.
+	--[[ Callback: Castbar:PostCastUpdate(unit)
+	Called after the element has been updated when a spell cast has been updated.
 
 	* self - the Castbar widget
-	* unit - unit that the update has been triggered (string)
-	* name - name of the delayed spell (string)
+	* unit - the unit that the update has been triggered (string)
 	--]]
-	if(element.PostCastDelayed) then
-		return element:PostCastDelayed(unit, name)
+	if(element.PostCastUpdate) then
+		return element:PostCastUpdate(unit)
 	end
 end
 
-local function UNIT_SPELLCAST_STOP(self, event, unit, castID)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+local function CastStop(self, event, unit, castID, spellID)
+	if(self.unit ~= unit) then return end
 
 	local element = self.Castbar
-	if(element.castID ~= castID) then
+	if(not element:IsShown() or element.castID ~= castID or element.spellID ~= spellID) then
 		return
 	end
 
-	element.casting = nil
-	element.notInterruptible = nil
+	resetAttributes(element)
 
-	--[[ Callback: Castbar:PostCastStop(unit)
-	Called after the element has been updated when a spell cast has finished.
+	--[[ Callback: Castbar:PostCastStop(unit, spellID)
+	Called after the element has been updated when a spell cast has stopped.
```

Comment:
Called after the element has been updated when a spell cast**/channel** has stopped.